File: Problem14/Problem14.c
//
// Created by georg on 17.04.17.
//

#include <stdio.h>
#include <time.h>
#include <malloc.h>

typedef int bool;
enum { false, true };

int compute_next_collatz_element(int n){
    if (n % 2 == 0){
        return n/2;
    }
    return 3*n+1;

}

int compute_collatz_seq_length(int startval){
    int counter = 0;
    int collatz_val = startval;
    while(collatz_val != 1){
        counter += 1;
        collatz_val = compute_next_collatz_element(collatz_val);
    }
    //printf("Length for %i: %i \n", startval, counter);
    return counter;
}

int compute_collatz_seq_length_with_memory(int startval, int* sequence_storage){
    int counter = 0;
    int collatz_val = startval;

    while(collatz_val != 1){
        counter += 1;
        collatz_val = compute_next_collatz_element(collatz_val);
        if(collatz_val < startval){
            return sequence_storage[collatz_val]+counter;
        }
    }
    //printf("Length for %i: %i \n", startval, counter);
    return counter;
}

int find_largest_collatz_seq_in_range(int threshhold){
    int largest_seq_length = 1;
    int temp;
    int largest_input_val = 1;

    for (int i = 1; i <= threshhold; i++) {
        temp = compute_collatz_seq_length(i);
        if (temp > largest_seq_length) {
            largest_seq_length = temp;
            largest_input_val = i;
        }
    }

    return largest_input_val;
}

int find_largest_collatz_seq_in_range_wth_storage(int N) {
    int largest_seq_length = 1;
    int temp;
    int largest_input_val = 1;
    int* stored_collatzlengths = malloc(N*sizeof(int));
    //int stored_collatzlengths[N];

    // get rid later
    for(int i=0;i<N;i++){
        stored_collatzlengths[i] = -1;
        // printf("Array[%i]=%i \n", i, stored_collatzlengths[i]);
    }


    for (int i = 1; i <= N; i++) {
        temp = compute_collatz_seq_length_with_memory(i, stored_collatzlengths);
        stored_collatzlengths[i] = temp;

        if (temp > largest_seq_length) {
            largest_seq_length = temp;
            largest_input_val = i;
        }
    }

    /*
    for (int i = 1; i <= threshhold; i++) {
        printf("Array[%i] = %i \n", i-1, stored_collatzlengths[i-1]);
    }
    */

    return largest_input_val;

}


void main(){
    int N = 100000;
    int res;
    bool use_memory = false;
    clock_t begin = clock();
    if(use_memory==true) {
        res = find_largest_collatz_seq_in_range_wth_storage(N);
    }
    else{
        res = find_largest_collatz_seq_in_range(N);
    }
    clock_t end = clock();
    double time_spent = (double)(end - begin) / CLOCKS_PER_SEC;

    printf("Largest Collatz Sequence under %i: %i \n", N, res);
    printf("Wall Time: %f \n", time_spent);
}

File: Problem14/Problem14.jl
function compute_next_collatz_element(elem::Int64)
  if (elem % 2 == 0)
    # use Integer division
    return div(elem, 2)
  else
    return 3*elem+1
  end
end

function compute_collatz_seq_length_with_memory(startval::Int64, sequence_storage::Array{Int64, 1})
    """
    Computes the length of the collatz sequence by reading the length of the
    Collatz sequence by being able to read from memory if an already computed
    Collatz sequence length can be used.
    """
    counter = 0
    collatz_val = startval

    while collatz_val != 1
        counter += 1
        collatz_val = compute_next_collatz_element(collatz_val)
        if collatz_val < startval
            # as we compute the Collatz sequence length in increasing order,
            # we hit an already computed Collatz sequence length from the
            # previous value and add how many steps it took to get there
            result = counter + sequence_storage[collatz_val]
            sequence_storage[startval] = result
            return result
        end
    end
    return counter
end


function find_largest_collatz_seq_in_range_wth_storage(N::Int64)
    """
    This finds the largest Collatz sequence length in the range from 1-N.
    """
    largest_seq_length = 1
    largest_input_val = 1
    stored_collatzlengths = Array{Int64}(N+1)
    stored_collatzlengths[1] = 1

    for i in 2:N
        seq_length = compute_collatz_seq_length_with_memory(i, stored_collatzlengths)

        if seq_length > largest_seq_length
            largest_seq_length = seq_length
            largest_input_val = i
        end
    end

    return largest_input_val

end

N=1000000
println("N = ", N)
@time res = find_largest_collatz_seq_in_range_wth_storage(N)
println("Result = ", res)
